File: sdk-go/filesystem.go
package workspace

import (
	"context"
	"io"

	pb "github.com/OpenElevo/ElevoSandbox/sdk-go/proto/workspace/v1"
)

// FileSystemService provides low-level filesystem operations via gRPC
// This is primarily used for FUSE mounting
type FileSystemService struct {
	client *Client
}

// DownloadBinary downloads a binary file from the server
func (f *FileSystemService) DownloadBinary(ctx context.Context, name, platform, arch string) ([]byte, error) {
	req := &pb.DownloadBinaryRequest{
		Name:     name,
		Platform: platform,
		Arch:     arch,
	}

	stream, err := f.client.fsClient.DownloadBinary(f.client.withAuth(ctx), req)
	if err != nil {
		return nil, convertGrpcError(err)
	}

	var data []byte
	for {
		resp, err := stream.Recv()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, convertGrpcError(err)
		}
		data = append(data, resp.Chunk...)
	}

	return data, nil
}

File: sdk-typescript/src/services/filesystem.ts
/**
 * FileSystem service for low-level filesystem operations via gRPC
 * Primarily used for FUSE mounting and binary downloads.
 */

import * as grpc from '@grpc/grpc-js';
import { FileSystemServiceClient, createMetadata } from '../grpc';
import { convertGrpcError } from '../errors';

/**
 * Service for low-level filesystem operations (FUSE support)
 */
export class FileSystemService {
  constructor(
    private readonly client: FileSystemServiceClient,
    private readonly apiKey?: string
  ) {}

  private metadata(): grpc.Metadata {
    return createMetadata(this.apiKey);
  }

  /**
   * Download a binary file from the server (e.g., workspace-fuse)
   */
  async downloadBinary(name: string, platform: string, arch: string): Promise<Uint8Array> {
    const request = { name, platform, arch };

    return new Promise((resolve, reject) => {
      const stream = this.client.downloadBinary(request, this.metadata());
      const chunks: Buffer[] = [];

      stream.on('data', (response: any) => {
        if (response.chunk) {
          chunks.push(Buffer.from(response.chunk));
        }
      });

      stream.on('end', () => {
        resolve(new Uint8Array(Buffer.concat(chunks)));
      });

      stream.on('error', (err: grpc.ServiceError) => {
        reject(convertGrpcError(err));
      });
    });
  }
}
